l1-contracts/contracts/bridge/L1NativeTokenVault.sol

```diff
@@ -96,38 +100,54 @@ contract L1NativeTokenVault is
-                balanceAfter -
-                balanceBefore;
         } else {
             uint256 balanceBefore = IERC20(_token).balanceOf(address(this));
-            uint256 sharedBridgeChainBalance = chainBalance[_targetChainId][_token];
+            uint256 sharedBridgeChainBalance = IERC20(_token).balanceOf(address(L1_SHARED_BRIDGE));
             require(sharedBridgeChainBalance > 0, "NTV: 0 amount to transfer");
-            L1_SHARED_BRIDGE.transferTokenToNTV(_targetChainId, _token);
+            L1_SHARED_BRIDGE.transferTokenToNTV(_token);
             uint256 balanceAfter = IERC20(_token).balanceOf(address(this));
             require(balanceAfter - balanceBefore >= sharedBridgeChainBalance, "NTV: wrong amount transferred");
-            chainBalance[_targetChainId][_token] = chainBalance[_targetChainId][_token] + sharedBridgeChainBalance;
         }
     }
 
-    /// @dev transfer tokens from legacy erc20 bridge or mailbox and set chainBalance as part of migration process.
-    /// @dev Unlike `transferFundsFromLegacy` is provides a concrete limit on the gas used for the transfer and even if it will fail, it will not revert the whole transaction.
-    function safeTransferFundsFromSharedBridge(
+    /// @dev Set chain token balance as part of migration process.
+    /// @param _token The address of token to be transferred (address(1) for ether and contract address for ERC20).
+    /// @param _targetChainId The chain ID of the corresponding hyperchain.
+    function transferBalancesFromSharedBridge(address _token, uint256 _targetChainId) external onlySelf {
+        require(
+            chainBalanceMigrated[_targetChainId][_token] == false,
+            "NTV: chain balance for the token already migrated"
+        );
+        uint256 sharedBridgeChainBalance = L1_SHARED_BRIDGE.chainBalance(_targetChainId, _token);
+        chainBalance[_targetChainId][_token] = chainBalance[_targetChainId][_token] + sharedBridgeChainBalance;
+        chainBalanceMigrated[_targetChainId][_token] == true;
+    }
+
+    /// @dev Transfer tokens from shared bridge as part of migration process.
+    /// @dev Unlike `transferFundsFromSharedBridge` is provides a concrete limit on the gas used for the transfer and even if it will fail, it will not revert the whole transaction.
+    function safeTransferFundsFromSharedBridge(address _token, uint256 _gasPerToken) external {
+        try this.transferFundsFromSharedBridge{gas: _gasPerToken}(_token) {} catch {
+            // A reasonable amount of gas will be provided to transfer the token.
+            // If the transfer fails, we don't want to revert the whole transaction.
+        }
+    }
+
+    /// @dev Set chain token balance as part of migration process.
+    /// @dev Unlike `transferBalancesFromSharedBridge` is provides a concrete limit on the gas used for the transfer and even if it will fail, it will not revert the whole transaction.
+    function safeTransferBalancesFromSharedBridge(
```

Comment:
These two functions should be the same. We should be transferring the funds I think at the same time we are transferring the balances